[Cocoa] Migrate to AVAssetWriter segment-based delegate callbacks

jernoble opened the pull request:
#### bb4974477bf7528d81299d2195f18c451ffe0e13
<pre>
[Cocoa] Migrate to AVAssetWriter segment-based delegate callbacks
<a href="https://bugs.webkit.org/show_bug.cgi?id=279989">https://bugs.webkit.org/show_bug.cgi?id=279989</a>
<a href="https://rdar.apple.com/136298814">rdar://136298814</a>

Reviewed by Eric Carlson.

Migrate to the -assetWriter:didProduceFragmentedHeaderData: delegate callback, away from
the deprecated fragment-based delegate callback.

This new callback has a stricter requirement about starting new segments with a keyframe.
To achieve this, modify VideoSampleBufferCompressor::flushInternal() to set a flag which
will cause the next enqueued sample to be a keyframe.

Drive-by fix: initialize the Options to be passed to MediaRecorderPrivate with the current
m_options before overriding the audio and video bitrates.

* Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
(WebCore::MediaRecorder::startRecording):
* Source/WebCore/PAL/pal/cf/VideoToolboxSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/VideoToolboxSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm:
(-[WebAVAssetWriterDelegate assetWriter:didOutputSegmentData:segmentType:]):
(WebCore::MediaRecorderPrivateWriter::initialize):
(WebCore::MediaRecorderPrivateWriter::stopRecording):
(WebCore::MediaRecorderPrivateWriter::fetchData):
(-[WebAVAssetWriterDelegate assetWriter:didProduceFragmentedHeaderData:]): Deleted.
(-[WebAVAssetWriterDelegate assetWriter:didProduceFragmentedMediaData:fragmentedMediaDataReport:]): Deleted.
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.mm:
(WebCore::VideoSampleBufferCompressor::flushInternal):
(WebCore::forceKeyFrameDictionary):
(WebCore::VideoSampleBufferCompressor::processSampleBuffer):

Canonical link: <a href="https://commits.webkit.org/283956@main">https://commits.webkit.org/283956@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9648ded2fe34d9b22db68372fe025fb04d5e142c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67843 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47225 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71899 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/18984 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69961 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55024 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18791 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54272 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12685 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70910 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43294 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58662 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34739 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39969 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16072 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17342 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61932 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16415 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73596 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11806 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15701 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61723 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11843 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58736 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15085 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9625 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3242 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43035 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/44109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45295 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43847 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/Modules/webaudio/AudioWorkletGlobalScope.cpp:
```diff
@@ -144,6 +144,10 @@ RefPtr<AudioWorkletProcessor> AudioWorkletGlobalScope::createProcessor(const Str
+    ASSERT(jsConstructor);
+    if (!jsConstructor)
+        return nullptr;
+
```
Is this change meant to be part of this patch?

jernoble replied:
Nope, this was part of a different commit that got swept up in my initial PR.

eric-carlson commented on Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.mm:
```diff
@@ -203,7 +216,11 @@ static inline OSStatus setCompressionSessionProperty(VTCompressionSessionRef vtS
-    auto error = PAL::VTCompressionSessionEncodeFrame(m_vtSession.get(), imageBuffer, presentationTimeStamp, duration, NULL, this, NULL);
+    auto error = PAL::VTCompressionSessionEncodeFrame(m_vtSession.get(), imageBuffer, presentationTimeStamp, duration, m_needsKeyframe ? forceKeyFrameDictionary() : NULL, this, NULL);
+
+    if (m_needsKeyframe)
```
It may be faster to avoid the test and just always clear the bool

jernoble committed:
[Cocoa] Migrate to AVAssetWriter segment-based delegate callbacks
https://bugs.webkit.org/show_bug.cgi?id=279989
rdar://136298814

Reviewed by Eric Carlson.

Migrate to the -assetWriter:didProduceFragmentedHeaderData: delegate callback, away from
the deprecated fragment-based delegate callback.

This new callback has a stricter requirement about starting new segments with a keyframe.
To achieve this, modify VideoSampleBufferCompressor::flushInternal() to set a flag which
will cause the next enqueued sample to be a keyframe.

Drive-by fix: initialize the Options to be passed to MediaRecorderPrivate with the current
m_options before overriding the audio and video bitrates.

* Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
(WebCore::MediaRecorder::startRecording):
* Source/WebCore/PAL/pal/cf/VideoToolboxSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/VideoToolboxSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm:
(-[WebAVAssetWriterDelegate assetWriter:didOutputSegmentData:segmentType:]):
(WebCore::MediaRecorderPrivateWriter::initialize):
(WebCore::MediaRecorderPrivateWriter::stopRecording):
(WebCore::MediaRecorderPrivateWriter::fetchData):
(-[WebAVAssetWriterDelegate assetWriter:didProduceFragmentedHeaderData:]): Deleted.
(-[WebAVAssetWriterDelegate assetWriter:didProduceFragmentedMediaData:fragmentedMediaDataReport:]): Deleted.
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.mm:
(WebCore::VideoSampleBufferCompressor::flushInternal):
(WebCore::forceKeyFrameDictionary):
(WebCore::VideoSampleBufferCompressor::processSampleBuffer):

Canonical link: https://commits.webkit.org/283956@main
